kernel/of/fdt_mem.c

```diff
@@ -0,0 +1,192 @@
+            return (size_t)0;
+        }
+        else {
+            // Need to trim before
+            size = (size_t)((start + size) - (void*)__kernel_end);
+            start = (void*)__kernel_end;
+        }
+    }
+    else {
+        // No addresses lower than the end of the kernel (so no overlap)
+    }
+
+    // Handle the reserve entries in the FDT
+    struct fdt_reserve_entry* entry = fdt_reserve_entry_begin(fdt);
+    while (entry != NULL) {
+        void* resv_start = fdt_reserve_entry_address(entry);
+        size_t resv_size = fdt_reserve_entry_size(entry);
+
+        if (trim_or_split_around_reserved(
+                &start, &size, &split_start, &split_size, resv_start, resv_size
+            )) {
+            // We split into 2 regions
+            return try_boot_free_region(fdt, start, size)
+                   + try_boot_free_region(fdt, split_start, split_size);
+        }
+        else {
+            if (size == 0) {
+                // We got fully trimmed
+                return (size_t)0;
+            }
+            else {
+                // We're still one contiguous region (continue)
+            }
+        }
+
+        entry = fdt_next_reserve_entry(fdt, entry);
+    }
+
+    // Make sure we don't free the FDT blob itself
+    void* fdt_resv_start = (void*)fdt;
+    size_t fdt_resv_size = fdt_size(fdt);
+
+    if (trim_or_split_around_reserved(
+            &start, &size, &split_start, &split_size, fdt_resv_start, fdt_resv_size
+        )) {
+        // We split into 2 regions
+        return try_boot_free_region(fdt, start, size)
+               + try_boot_free_region(fdt, split_start, split_size);
+    }
+    else {
+        if (size == 0) {
+            // We got fully trimmed
+            return (size_t)0;
+        }
+        else {
+            // We're still one contiguous region (continue)
+        }
+    }
+
+    // We made it through (possibly shrunk)
+    if (boot_free(start, size) != KERR_SUCCESS) {
+        // We couldn't free for some reason
+        return (size_t)0;
+    }
+    return size;
+}
+
+/*
+ * Use the FDT to hand off free memory regions to the "boot" memory allocator
+ */
+kerror_t
+fdt_boot_mem_init(struct fdt* fdt)
+{
+    struct fdt_node* mem_node = fdt_find_node_by_device_type(fdt, NULL, "memory");
+
+    size_t total_mem_seen = 0;
+    size_t total_mem_freed = 0;
+
+    while (mem_node != NULL) {
+        // TODO: Actually free memory
+        size_t num_register_blocks = fdt_node_num_register_blocks(fdt, mem_node);
+
+        for (size_t i = 0; i < num_register_blocks; i++) {
+            void* address;
+            size_t size;
+
+            fdt_node_get_register_block(fdt, mem_node, 0, &address, &size);
+            total_mem_seen += size;
+
+            printk(
+                "memory_node: (%s)[%c] {addr = %p, size = %p}\n",
+                fdt_node_name(mem_node),
+                '0' + (unsigned char)i,
+                (void*)address,
+                (void*)(uintptr_t)size
```

Comment:
```suggestion
                (void*)size
```

we are on C23 and using an intelligent compiler, we don't need to do these old hacks

Reply:
"We're using an intelligent compiler" is the same line of thought as people misusing `int` and `long`, I'm not certain this cast is needed but it doesn't hurt anything so I'm leaving it.

Reply:
It makes readability harder. There's no point to use two casts when one works.

Reply:
actually never mind keep it its good

https://stackoverflow.com/a/34291521